### PR TITLE
Change "full" depth to average depth

### PR DESCRIPTION
**Pull request by dubslow**
Since there remain major disagreements about the performance of recalculating reused-tree depths outside of the Search code, this patch instead leaves the updated depth inside the Node structure, as currently exists in master. Even so, it's still technically less cpu cycles, although probably trivially so. It is also +2 bytes in Node, but it seems that these +2 bytes are already wasted by alignment, so at least on g++/x86 compiles, this is memory neutral. With this patch: `Move size 2 for estimated Edge size 6 vs actual Edge size 8 Node size 72 for total estimated bytes/Node of 352` and on master: `Move size 2 for estimated Edge size 6 vs actual Edge size 8 Node size 72 for total estimated bytes/Node of 352`. I'm pretty sure it didn't used to be this way until the introduction of the cached TotalVisitedPolicy thing, I guess that extra float is causing the alignment problem. (And yes, I double-triple checked I was actually using master when running the comparison.)

So it seems that this patch is strictly better than status quo, being identical memory and trivially less cpu, and also having the benefit that "full" depth is replaced with something much more [useful](https://github.com/LeelaChessZero/lc0/pull/124#issue-198461389) yet which also satisfies my "one move analysis, one depth" [criterion](https://github.com/LeelaChessZero/lc0/issues/86#issuecomment-408179482). (In particular, full depth is bugged with having sometimes depth > 1000 output, but most importantly, it always shows the ~same value for nearly all searches of any sort, including trees that are order of magnitudes different in size, or trees which are the same size but wildly different shape. Average depth fixes all these problems.)

Although I haven't had time to investigate further about the performance of my old PR #160, which I still provisionally believe to be superior pending further investigation (several bytes fewer memory, less total cpu time, less cpu time in search code), I'm willing to close that one if this one is approved.

### CODE REVIEW

**Comment by dubslow** [comment]
Actually, SunnyWar just provided an analysis of `Node`'s layout and alignment, and the biggest offender is that the `bool is_terminal_` is costing 8 of those 72 bytes, though I think this PR changes it to the bool costing 4 bytes and `uint32_t cumulative_depth_` costing 6 bytes, though there's other alignment penalties that can probably be optimized as well. (We could even shrink `visited_policy_` to 2 bytes like in the policy-shrink PR if that would help.)

**Comment by SunnyWar** [comment]
@dubslow I've been poking around at some of the other objects that appear to be used a lot.  This one seems unusually large and I see several places where it is copied by value.

1>class lczero::MoveExecution	size(64):
1>	+---
1> 0	| Move move
1>  	| <alignment member> (size=6)
1> 8	| ChessBoard board
1>56	| reset_50_moves
1>  	| <alignment member> (size=7)
1>	+---

Rearranging saves a bit of space:

1>class lczero::MoveExecution	size(56):
1>	+---
1> 0	| ChessBoard board
1>48	| Move move
1>50	| reset_50_moves
1>  	| <alignment member> (size=5)
1>	+---

File: src/mcts/node.h [hunk 1]
```diff
@@ -177,16 +174,25 @@ class Node {
   // * N-in-flight (-=1)
   void FinalizeScoreUpdate(float v);
 
-  // Updates max depth, if new depth is larger.
-  void UpdateMaxDepth(int depth);
+  // Depth tracking and retrieval functions.
+  void UpdateMaxDepth(const uint16_t& depth) {
```
Comment (SunnyWar) [inline src/mcts/node.h]: Why pass by reference? As is it compiles to void UpdateAverageDepth(unsigned short const & __ptr64)
versus void UpdateAverageDepth(const uint16_t depth) which compile to void UpdateAverageDepth(unsigned short)

File: src/mcts/node.h [hunk 2]
```diff
@@ -177,16 +174,25 @@ class Node {
...
+    if (depth > max_depth_) max_depth_ = depth;
+  }
+  void UpdateAverageDepth(const uint16_t& depth) {
+    // Denominator should be total+1, but we exclude the root from the average.
+    average_depth_ += (depth - average_depth_) / n_;
+  }
```
Comment (SunnyWar) [inline src/mcts/node.h]: Why pass by reference? As is it compiles to void UpdateAverageDepth(unsigned short const & __ptr64)
versus void UpdateAverageDepth(const uint16_t depth) which compile to void UpdateAverageDepth(unsigned short)

**Comment by dubslow** [comment]
In support of this:

depth_poll

**Comment by mooskagh** [comment]
Closing for now.
But after tree detaching is implemented, there will be no depth reporting (initially) and we'll have a new round of depth discussion with fresh strength. :)

**Pull request closed without merging**